src/lib/tls/tls_messages.h

```diff
@@ -118,69 +199,77 @@ class BOTAN_UNSTABLE_API Client_Hello final : public Handshake_Message
 
       std::vector<uint8_t> session_ticket() const;
 
-      bool supports_alpn() const;
-
       bool supports_extended_master_secret() const;
 
       bool supports_cert_status_message() const;
 
       bool supports_encrypt_then_mac() const;
 
-      bool sent_signature_algorithms() const;
+      void update_hello_cookie(const Hello_Verify_Request& hello_verify);
+   };
 
-      std::vector<std::string> next_protocols() const;
+/**
+* Server Hello Message
+*/
+class BOTAN_UNSTABLE_API Server_Hello : public Handshake_Message
+   {
+   public:
+      std::vector<uint8_t> serialize() const override;
 
-      std::vector<uint16_t> srtp_profiles() const;
+      Handshake_Type type() const override;
 
-      void update_hello_cookie(const Hello_Verify_Request& hello_verify);
+      // methods available in both subclasses' interface
+      uint16_t ciphersuite() const;
+      const Extensions& extensions() const;
+      const std::vector<uint8_t>& session_id() const;
 
-      const std::vector<uint8_t>& cookie() const { return m_hello_cookie; }
+      virtual Protocol_Version selected_version() const = 0;
 
-      std::vector<uint8_t> cookie_input_data() const;
+   protected:
+      /**
+       * Version-agnostic internal server hello data container that allows
+       * parsing Server_Hello messages without prior knowledge of the contained
+       * protocol version.
+       */
+      class Internal
+         {
+         public:
+            Internal(const std::vector<uint8_t>& buf);
 
-      std::set<Handshake_Extension_Type> extension_types() const
-         { return m_extensions.extension_types(); }
-
-      const Extensions& extensions() const { return m_extensions; }
-
-      Client_Hello(Handshake_IO& io,
-                   Handshake_Hash& hash,
-                   const Policy& policy,
-                   Callbacks& cb,
-                   RandomNumberGenerator& rng,
-                   const std::vector<uint8_t>& reneg_info,
-                   const Client_Hello::Settings& client_settings,
-                   const std::vector<std::string>& next_protocols);
-
-      Client_Hello(Handshake_IO& io,
-                   Handshake_Hash& hash,
-                   const Policy& policy,
-                   Callbacks& cb,
-                   RandomNumberGenerator& rng,
-                   const std::vector<uint8_t>& reneg_info,
-                   const Session& resumed_session,
-                   const std::vector<std::string>& next_protocols);
+            Internal(Protocol_Version legacy_version,
+                     std::vector<uint8_t> session_id,
+                     std::vector<uint8_t> random,
+                     const uint16_t ciphersuite,
+                     const uint8_t comp_method);
 
-      explicit Client_Hello(const std::vector<uint8_t>& buf);
+            Protocol_Version version() const;
 
-   private:
-      std::vector<uint8_t> serialize() const override;
+         public:
```

Comment:
`Server_Hello::Internal` will act as a version-agnostic parser and "data holder" after TLS 1.3 is added. It encapsulate the parsing code for `Server_Hello_12`, `Server_Hello_13` and `Hello_Retry_Request`. Those all share the same wire data model, hence the indirection.

We deemed it unnecessary to hide those members from the anyway tightly coupled concrete server hello type. Especially since `Server_Hello::Internal` isn't intended to leak out of the various server hello objects.

That said: Probably we should not define it in the header then. I'll look into it.

Reply:
In the hope of a smoother rebase after this is merged I applied the adaption on `dev/tls-13` instead of in this intermediate PR. Please see this patch: https://github.com/randombit/botan/pull/2922/commits/b2fcaa8a66348cfa6c77403cb61608fa22402fe1#diff-c6b0802c7b3ef40a3ef569f9072ebfc9452dc8b019e94e1e9c3449754019d088